### ISSUE TITLE
Graph column width sometimes appears too narrow

### ISSUE THREAD

**Issue by drewnoakes** [gitextensions/gitextensions#5167]
**Do you want to request a *feature* or report a *bug*?**

Bug, following #5087.

**What is the current behavior?**

Graph can appear too narrow, with width for only a single lane.

![image](https://user-images.githubusercontent.com/350947/42738567-b086b406-887d-11e8-956c-68e5c4000266.png)

Scrolling triggers the correct sizing:

![image](https://user-images.githubusercontent.com/350947/42738571-bd4b15f6-887d-11e8-87c9-7e2b8cfc5590.png)

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

1. Ensure object tree visible
1. Refresh revisions
1. Click a branch in the object tree 

@mstv has [another reliable method](https://github.com/gitextensions/gitextensions/pull/5087#issuecomment-401893955):

1. open the `Commit` dialog
1. maximize the `Commit` dialog window (important!)
1. check `Amend Commit`
1. click `Commit` (without changing anything)
1. confirm to rewrite the history
1. repeat at most once from 1.

**What is the expected behavior?**

Graph always appears in the correct width.

**Environment you encounter the issue:**

- GitExtensions version: v3-ux branch from #5087.

**Comment by RussKie** [comment]
I observe it in 8f80b287cfcf6731161775861b7bbaeb2ea2bd06.
A repro is pretty consistent - all I have to do is _fetch all_ and voila

![image](https://user-images.githubusercontent.com/4403806/43396977-b021e2f4-9446-11e8-8a24-8ff0d4372dfd.png)

**Comment by drewnoakes** [comment]
One day soon I will sit down and review all these graph glitches in one go.

**Comment by RussKie** [comment]
I've done some investigation and I'm starting to overheat looking at the code.

Here's what I concluded so far:
* all lines are always drawn properly if the top commit is selected (I couldn't fault it)
* if a commit beyond clip area is selected (e.g. one has to scroll to select the commit) then it can be one of three outcomes upon refresh:
  - all lanes are sized and drawn correctly (intermittent)
  - some lanes are not drawn, the column width is somewhat correct
  - the column is squashed

Here's a screenshot of a repo scrolled by 10 rows, the top row has index 10 (i.e. the 11th row, 2nd from the bottom, if scrolled to the top)
![image](https://user-images.githubusercontent.com/4403806/44341136-633b9b80-a4ca-11e8-8cd5-453b3ff71af4.png)

I have sprinkled the code with some debug statements:

```
> PaintGraphCell: 0, {X=0,Y=0,Width=140,Height=52}
  > 1: 0 - 1, 0:1
  > DrawVisibleGraph: 0 - 1
    > Creating CreateRectangle
  > CreateRectangle: 0 - 1, {X=0,Y=0,Width=140,Height=52}
> PaintGraphCell: 1, {X=0,Y=52,Width=140,Height=52}
  > 1: 1 - 2, 0:2
  > DrawVisibleGraph: 1 - 2
    > Creating CreateRectangle
  > CreateRectangle: 1 - 2, {X=0,Y=52,Width=140,Height=52}
> PaintGraphCell: 2, {X=0,Y=104,Width=140,Height=52}
  > 1: 2 - 3, 0:3
  > DrawVisibleGraph: 2 - 3
    > Creating CreateRectangle
  > CreateRectangle: 2 - 3, {X=0,Y=104,Width=140,Height=52}
> PaintGraphCell: 3, {X=0,Y=156,Width=140,Height=52}
  > 1: 3 - 4, 0:4
  > DrawVisibleGraph: 3 - 4
    > Creating CreateRectangle
  > CreateRectangle: 3 - 4, {X=0,Y=156,Width=140,Height=52}
> PaintGraphCell: 4, {X=0,Y=208,Width=140,Height=52}
  > 1: 4 - 5, 0:5
  > DrawVisibleGraph: 4 - 5
    > Creating CreateRectangle
  > CreateRectangle: 4 - 5, {X=0,Y=208,Width=140,Height=52}
> PaintGraphCell: 5, {X=0,Y=260,Width=140,Height=52}
  > 1: 5 - 6, 0:6
  > DrawVisibleGraph: 5 - 6
    > Creating CreateRectangle
  > CreateRectangle: 5 - 6, {X=0,Y=260,Width=140,Height=52}
> PaintGraphCell: 6, {X=0,Y=312,Width=140,Height=52}
  > 1: 6 - 7, 0:7
  > DrawVisibleGraph: 6 - 7
    > Creating CreateRectangle
  > CreateRectangle: 6 - 7, {X=0,Y=312,Width=140,Height=52}
> PaintGraphCell: 7, {X=0,Y=364,Width=140,Height=52}
  > 1: 7 - 8, 0:8
  > DrawVisibleGraph: 7 - 8
    > Creating CreateRectangle
  > CreateRectangle: 7 - 8, {X=0,Y=364,Width=140,Height=52}
> PaintGraphCell: 8, {X=0,Y=416,Width=140,Height=52}
  > 1: 8 - 9, 0:9
  > DrawVisibleGraph: 8 - 9
    > Creating CreateRectangle
  > CreateRectangle: 8 - 9, {X=0,Y=416,Width=140,Height=52}
> PaintGraphCell: 9, {X=0,Y=468,Width=140,Height=52}
  > 1: 9 - 10, 0:10
  > DrawVisibleGraph: 9 - 10
    > Creating CreateRectangle
  > CreateRectangle: 9 - 10, {X=0,Y=468,Width=140,Height=52}
> PaintGraphCell: 10, {X=0,Y=520,Width=140,Height=52}
  > 1: 10 - 11, 0:11
  > DrawVisibleGraph: 10 - 11
    > Creating CreateRectangle
  > CreateRectangle: 10 - 11, {X=0,Y=520,Width=140,Height=52}
> PaintGraphCell: 11, {X=0,Y=572,Width=140,Height=52}
  > 1: 11 - 12, 0:12
  > DrawVisibleGraph: 11 - 12
    > Creating CreateRectangle
  > CreateRectangle: 11 - 12, {X=0,Y=572,Width=140,Height=52}
> PaintGraphCell: 12, {X=0,Y=624,Width=140,Height=52}
  > 1: 12 - 13, 0:13
  > DrawVisibleGraph: 12 - 13
    > Failed to DrawVisibleGraph
	
END 1 ----------------------------------------------------------------------------------------
	
> PaintGraphCell: 0, {X=0,Y=0,Width=140,Height=52}
  > 1: 0 - 1, 0:1
  > DrawVisibleGraph: 0 - 1
    > Creating CreateRectangle
  > CreateRectangle: 0 - 1, {X=0,Y=0,Width=140,Height=52}
> PaintGraphCell: 1, {X=0,Y=52,Width=140,Height=52}
  > 1: 1 - 2, 0:2
  > DrawVisibleGraph: 1 - 2
    > Creating CreateRectangle
  > CreateRectangle: 1 - 2, {X=0,Y=52,Width=140,Height=52}
> PaintGraphCell: 2, {X=0,Y=104,Width=140,Height=52}
  > 1: 2 - 3, 0:3
  > DrawVisibleGraph: 2 - 3
    > Creating CreateRectangle
  > CreateRectangle: 2 - 3, {X=0,Y=104,Width=140,Height=52}
> PaintGraphCell: 3, {X=0,Y=156,Width=140,Height=52}
  > 1: 3 - 4, 0:4
  > DrawVisibleGraph: 3 - 4
    > Creating CreateRectangle
  > CreateRectangle: 3 - 4, {X=0,Y=156,Width=140,Height=52}
> PaintGraphCell: 4, {X=0,Y=208,Width=140,Height=52}
  > 1: 4 - 5, 0:5
  > DrawVisibleGraph: 4 - 5
    > Creating CreateRectangle
  > CreateRectangle: 4 - 5, {X=0,Y=208,Width=140,Height=52}
> PaintGraphCell: 5, {X=0,Y=260,Width=140,Height=52}
  > 1: 5 - 6, 0:6
  > DrawVisibleGraph: 5 - 6
    > Creating CreateRectangle
  > CreateRectangle: 5 - 6, {X=0,Y=260,Width=140,Height=52}
> PaintGraphCell: 6, {X=0,Y=312,Width=140,Height=52}
  > 1: 6 - 7, 0:7
  > DrawVisibleGraph: 6 - 7
    > Creating CreateRectangle
  > CreateRectangle: 6 - 7, {X=0,Y=312,Width=140,Height=52}
> PaintGraphCell: 7, {X=0,Y=364,Width=140,Height=52}
  > 1: 7 - 8, 0:8
  > DrawVisibleGraph: 7 - 8
    > Creating CreateRectangle
  > CreateRectangle: 7 - 8, {X=0,Y=364,Width=140,Height=52}
> PaintGraphCell: 8, {X=0,Y=416,Width=140,Height=52}
  > 1: 8 - 9, 0:9
  > DrawVisibleGraph: 8 - 9
    > Creating CreateRectangle
  > CreateRectangle: 8 - 9, {X=0,Y=416,Width=140,Height=52}
> PaintGraphCell: 9, {X=0,Y=468,Width=140,Height=52}
  > 1: 9 - 10, 0:10
  > DrawVisibleGraph: 9 - 10
    > Creating CreateRectangle
  > CreateRectangle: 9 - 10, {X=0,Y=468,Width=140,Height=52}
> PaintGraphCell: 10, {X=0,Y=520,Width=140,Height=52}
  > 1: 10 - 11, 0:11
  > DrawVisibleGraph: 10 - 11
    > Creating CreateRectangle
  > CreateRectangle: 10 - 11, {X=0,Y=520,Width=140,Height=52}
> PaintGraphCell: 11, {X=0,Y=572,Width=140,Height=52}
  > 1: 11 - 12, 0:12
  > DrawVisibleGraph: 11 - 12
    > Creating CreateRectangle
  > CreateRectangle: 11 - 12, {X=0,Y=572,Width=140,Height=52}
> PaintGraphCell: 12, {X=0,Y=624,Width=140,Height=52}
  > 1: 12 - 13, 0:13
  > DrawVisibleGraph: 12 - 13
    > Failed to DrawVisibleGraph
	
END 2 ----------------------------------------------------------------------------------------
	
> PaintGraphCell: 0, {X=0,Y=0,Width=140,Height=52}
  > 1: 0 - 1, 0:1
  > DrawVisibleGraph: 0 - 1
    > Creating CreateRectangle
  > CreateRectangle: 0 - 1, {X=0,Y=0,Width=140,Height=52}
> PaintGraphCell: 1, {X=0,Y=52,Width=140,Height=52}
  > 1: 1 - 2, 0:2
  > DrawVisibleGraph: 1 - 2
    > Creating CreateRectangle
  > CreateRectangle: 1 - 2, {X=0,Y=52,Width=140,Height=52}
> PaintGraphCell: 2, {X=0,Y=104,Width=140,Height=52}
  > 1: 2 - 3, 0:3
  > DrawVisibleGraph: 2 - 3
    > Creating CreateRectangle
  > CreateRectangle: 2 - 3, {X=0,Y=104,Width=140,Height=52}
> PaintGraphCell: 3, {X=0,Y=156,Width=140,Height=52}
  > 1: 3 - 4, 0:4
  > DrawVisibleGraph: 3 - 4
    > Creating CreateRectangle
  > CreateRectangle: 3 - 4, {X=0,Y=156,Width=140,Height=52}
> PaintGraphCell: 4, {X=0,Y=208,Width=140,Height=52}
  > 1: 4 - 5, 0:5
  > DrawVisibleGraph: 4 - 5
    > Creating CreateRectangle
  > CreateRectangle: 4 - 5, {X=0,Y=208,Width=140,Height=52}
> PaintGraphCell: 5, {X=0,Y=260,Width=140,Height=52}
  > 1: 5 - 6, 0:6
  > DrawVisibleGraph: 5 - 6
    > Creating CreateRectangle
  > CreateRectangle: 5 - 6, {X=0,Y=260,Width=140,Height=52}
> PaintGraphCell: 6, {X=0,Y=312,Width=140,Height=52}
  > 1: 6 - 7, 0:7
  > DrawVisibleGraph: 6 - 7
    > Creating CreateRectangle
  > CreateRectangle: 6 - 7, {X=0,Y=312,Width=140,Height=52}
> PaintGraphCell: 7, {X=0,Y=364,Width=140,Height=52}
  > 1: 7 - 8, 0:8
  > DrawVisibleGraph: 7 - 8
    > Creating CreateRectangle
  > CreateRectangle: 7 - 8, {X=0,Y=364,Width=140,Height=52}
> PaintGraphCell: 8, {X=0,Y=416,Width=140,Height=52}
  > 1: 8 - 9, 0:9
  > DrawVisibleGraph: 8 - 9
    > Creating CreateRectangle
  > CreateRectangle: 8 - 9, {X=0,Y=416,Width=140,Height=52}
> PaintGraphCell: 9, {X=0,Y=468,Width=140,Height=52}
  > 1: 9 - 10, 0:10
  > DrawVisibleGraph: 9 - 10
    > Creating CreateRectangle
  > CreateRectangle: 9 - 10, {X=0,Y=468,Width=140,Height=52}
> PaintGraphCell: 10, {X=0,Y=520,Width=140,Height=52}
  > 1: 10 - 11, 0:11
  > DrawVisibleGraph: 10 - 11
    > Creating CreateRectangle
  > CreateRectangle: 10 - 11, {X=0,Y=520,Width=140,Height=52}
> PaintGraphCell: 11, {X=0,Y=572,Width=140,Height=52}
  > 1: 11 - 12, 0:12
  > DrawVisibleGraph: 11 - 12
    > Creating CreateRectangle
  > CreateRectangle: 11 - 12, {X=0,Y=572,Width=140,Height=52}
> PaintGraphCell: 12, {X=0,Y=624,Width=140,Height=52}
  > 1: 12 - 13, 0:13
  > DrawVisibleGraph: 12 - 13
    > Failed to DrawVisibleGraph
	
END 3 ----------------------------------------------------------------------------------------
	
> PaintGraphCell: 0, {X=0,Y=0,Width=140,Height=52}
  > 1: 0 - 1, 0:1
  > DrawVisibleGraph: 0 - 1
    > Creating CreateRectangle
  > CreateRectangle: 0 - 1, {X=0,Y=0,Width=140,Height=52}
> PaintGraphCell: 1, {X=0,Y=52,Width=140,Height=52}
  > 1: 1 - 2, 0:2
  > DrawVisibleGraph: 1 - 2
    > Creating CreateRectangle
  > CreateRectangle: 1 - 2, {X=0,Y=52,Width=140,Height=52}
> PaintGraphCell: 2, {X=0,Y=104,Width=140,Height=52}
  > 1: 2 - 3, 0:3
  > DrawVisibleGraph: 2 - 3
    > Creating CreateRectangle
  > CreateRectangle: 2 - 3, {X=0,Y=104,Width=140,Height=52}
> PaintGraphCell: 3, {X=0,Y=156,Width=140,Height=52}
  > 1: 3 - 4, 0:4
  > DrawVisibleGraph: 3 - 4
    > Creating CreateRectangle
  > CreateRectangle: 3 - 4, {X=0,Y=156,Width=140,Height=52}
> PaintGraphCell: 4, {X=0,Y=208,Width=140,Height=52}
  > 1: 4 - 5, 0:5
  > DrawVisibleGraph: 4 - 5
    > Creating CreateRectangle
  > CreateRectangle: 4 - 5, {X=0,Y=208,Width=140,Height=52}
> PaintGraphCell: 5, {X=0,Y=260,Width=140,Height=52}
  > 1: 5 - 6, 0:6
  > DrawVisibleGraph: 5 - 6
    > Creating CreateRectangle
  > CreateRectangle: 5 - 6, {X=0,Y=260,Width=140,Height=52}
> PaintGraphCell: 6, {X=0,Y=312,Width=140,Height=52}
  > 1: 6 - 7, 0:7
  > DrawVisibleGraph: 6 - 7
    > Creating CreateRectangle
  > CreateRectangle: 6 - 7, {X=0,Y=312,Width=140,Height=52}
> PaintGraphCell: 7, {X=0,Y=364,Width=140,Height=52}
  > 1: 7 - 8, 0:8
  > DrawVisibleGraph: 7 - 8
    > Creating CreateRectangle
  > CreateRectangle: 7 - 8, {X=0,Y=364,Width=140,Height=52}
> PaintGraphCell: 8, {X=0,Y=416,Width=140,Height=52}
  > 1: 8 - 9, 0:9
  > DrawVisibleGraph: 8 - 9
    > Creating CreateRectangle
  > CreateRectangle: 8 - 9, {X=0,Y=416,Width=140,Height=52}
> PaintGraphCell: 9, {X=0,Y=468,Width=140,Height=52}
  > 1: 9 - 10, 0:10
  > DrawVisibleGraph: 9 - 10
    > Creating CreateRectangle
  > CreateRectangle: 9 - 10, {X=0,Y=468,Width=140,Height=52}
> PaintGraphCell: 10, {X=0,Y=520,Width=140,Height=52}
  > 1: 10 - 11, 0:11
  > DrawVisibleGraph: 10 - 11
    > Creating CreateRectangle
  > CreateRectangle: 10 - 11, {X=0,Y=520,Width=140,Height=52}
> PaintGraphCell: 11, {X=0,Y=572,Width=140,Height=52}
  > 1: 11 - 12, 0:12
  > DrawVisibleGraph: 11 - 12
    > Creating CreateRectangle
  > CreateRectangle: 11 - 12, {X=0,Y=572,Width=140,Height=52}
> PaintGraphCell: 12, {X=0,Y=624,Width=140,Height=52}
  > 1: 12 - 13, 0:13
  > DrawVisibleGraph: 12 - 13
    > Failed to DrawVisibleGraph
	
END 4 ----------------------------------------------------------------------------------------
	
> PaintGraphCell: 10, {X=0,Y=0,Width=108,Height=52}
  > 1: 10 - 11, 0:11
  > DrawVisibleGraph: 10 - 11
    > Creating CreateRectangle
  > CreateRectangle: 10 - 11, {X=0,Y=520,Width=108,Height=52}
> PaintGraphCell: 11, {X=0,Y=52,Width=108,Height=52}
  > 1: 11 - 12, 0:12
  > DrawVisibleGraph: 11 - 12
    > Creating CreateRectangle
  > CreateRectangle: 11 - 12, {X=0,Y=572,Width=108,Height=52}
> PaintGraphCell: 12, {X=0,Y=104,Width=108,Height=52}
  > 1: 12 - 13, 0:13
  > DrawVisibleGraph: 12 - 13
    > Failed to DrawVisibleGraph
> PaintGraphCell: 13, {X=0,Y=156,Width=108,Height=52}
  > 1: 13 - 14, 0:14
  > DrawVisibleGraph: 13 - 14
    > Failed to DrawVisibleGraph
> PaintGraphCell: 14, {X=0,Y=208,Width=108,Height=52}
  > 1: 14 - 15, 0:15
  > DrawVisibleGraph: 14 - 15
    > Failed to DrawVisibleGraph
> PaintGraphCell: 15, {X=0,Y=260,Width=108,Height=52}
  > 1: 15 - 16, 0:16
  > DrawVisibleGraph: 15 - 16
    > Failed to DrawVisibleGraph
> PaintGraphCell: 16, {X=0,Y=312,Width=108,Height=52}
  > 1: 16 - 17, 0:17
  > DrawVisibleGraph: 16 - 17
    > Failed to DrawVisibleGraph
> PaintGraphCell: 17, {X=0,Y=364,Width=108,Height=52}
  > 1: 17 - 18, 0:18
  > DrawVisibleGraph: 17 - 18
    > Failed to DrawVisibleGraph
> PaintGraphCell: 18, {X=0,Y=416,Width=108,Height=52}
  > 1: 18 - 19, 0:19
  > DrawVisibleGraph: 18 - 19
    > Failed to DrawVisibleGraph
> PaintGraphCell: 19, {X=0,Y=468,Width=108,Height=52}
  > 1: 19 - 20, 0:20
  > DrawVisibleGraph: 19 - 20
    > Failed to DrawVisibleGraph
> PaintGraphCell: 20, {X=0,Y=520,Width=108,Height=52}
  > 1: 20 - 21, 0:21
  > DrawVisibleGraph: 20 - 21
    > Failed to DrawVisibleGraph
> PaintGraphCell: 21, {X=0,Y=572,Width=108,Height=52}
  > 1: 21 - 22, 0:22
  > DrawVisibleGraph: 21 - 22
    > Failed to DrawVisibleGraph
> PaintGraphCell: 22, {X=0,Y=624,Width=108,Height=52}
  > 1: 22 - 23, 0:23
  > DrawVisibleGraph: 22 - 23
    > Failed to DrawVisibleGraph
	
END 1 ----------------------------------------------------------------------------------------
	
> PaintGraphCell: 10, {X=0,Y=0,Width=108,Height=52}
  > 1: 10 - 11, 0:11
  > DrawVisibleGraph: 10 - 11
    > Creating CreateRectangle
  > CreateRectangle: 10 - 11, {X=0,Y=520,Width=108,Height=52}
> PaintGraphCell: 11, {X=0,Y=52,Width=108,Height=52}
  > 1: 11 - 12, 0:12
  > DrawVisibleGraph: 11 - 12
    > Creating CreateRectangle
  > CreateRectangle: 11 - 12, {X=0,Y=572,Width=108,Height=52}
> PaintGraphCell: 12, {X=0,Y=104,Width=108,Height=52}
  > 1: 12 - 13, 0:13
  > DrawVisibleGraph: 12 - 13
    > Failed to DrawVisibleGraph
> PaintGraphCell: 13, {X=0,Y=156,Width=108,Height=52}
  > 1: 13 - 14, 0:14
  > DrawVisibleGraph: 13 - 14
    > Failed to DrawVisibleGraph
> PaintGraphCell: 14, {X=0,Y=208,Width=108,Height=52}
  > 1: 14 - 15, 0:15
  > DrawVisibleGraph: 14 - 15
    > Failed to DrawVisibleGraph
> PaintGraphCell: 15, {X=0,Y=260,Width=108,Height=52}
  > 1: 15 - 16, 0:16
  > DrawVisibleGraph: 15 - 16
    > Failed to DrawVisibleGraph
> PaintGraphCell: 16, {X=0,Y=312,Width=108,Height=52}
  > 1: 16 - 17, 0:17
  > DrawVisibleGraph: 16 - 17
    > Failed to DrawVisibleGraph
> PaintGraphCell: 17, {X=0,Y=364,Width=108,Height=52}
  > 1: 17 - 18, 0:18
  > DrawVisibleGraph: 17 - 18
    > Failed to DrawVisibleGraph
> PaintGraphCell: 18, {X=0,Y=416,Width=108,Height=52}
  > 1: 18 - 19, 0:19
  > DrawVisibleGraph: 18 - 19
    > Failed to DrawVisibleGraph
> PaintGraphCell: 19, {X=0,Y=468,Width=108,Height=52}
  > 1: 19 - 20, 0:20
  > DrawVisibleGraph: 19 - 20
    > Failed to DrawVisibleGraph
> PaintGraphCell: 20, {X=0,Y=520,Width=108,Height=52}
  > 1: 20 - 21, 0:21
  > DrawVisibleGraph: 20 - 21
    > Failed to DrawVisibleGraph
> PaintGraphCell: 21, {X=0,Y=572,Width=108,Height=52}
  > 1: 21 - 22, 0:22
  > DrawVisibleGraph: 21 - 22
    > Failed to DrawVisibleGraph
> PaintGraphCell: 22, {X=0,Y=624,Width=108,Height=52}
  > 1: 22 - 23, 0:23
  > DrawVisibleGraph: 22 - 23
    > Failed to DrawVisibleGraph
	
END 2 ----------------------------------------------------------------------------------------
	
> PaintGraphCell: 21, {X=0,Y=572,Width=108,Height=52}
  > 1: 21 - 22, 0:22
  > DrawVisibleGraph: 21 - 22
    > Creating CreateRectangle
  > CreateRectangle: 21 - 22, {X=0,Y=1092,Width=108,Height=52}
  
```

What's interesting is that rows 0-12 are painted 4(!) times, then rows 10-22 are painted twice (!), and then the 2nd last visible row gets repainted again. Better to say "attempted to get painted".

So far I was unable to figure out the root cause for "Failed to DrawVisibleGraph", which is ultimately a result of the first `return false` in `DrawItem(Graphics g, ILaneRow row)`.

Legend:
* "_Failed to DrawVisibleGraph_" is emitted when the code exits at `if (!DrawVisibleGraph())`.
* `1: 22 - 23, 0:23` is emitted in 
    ```
    if (newRows > 0) { 1: {start} - {end}, {_cacheHead}:{_cacheCount} } 
    else if (neededHeadAdjustment > 0) { 2: ditto } 
    else if (neededHeadAdjustment < 0) { 3: ditto }
    ```


On a side note, I'm starting to think that private methods are evil - they lead to the same hard-to-follow code, especially considering reuse of local variables across method boundaries...

**Comment by RussKie** [comment]
Further investigation updates.

It looks like whenever we refresh the graph it fails to correctly count visible lanes, the resultant `laneCount = 1`:

https://github.com/gitextensions/gitextensions/blob/33d3decce33069ee4d96dcaeecef09ad33683e97/GitUI/UserControls/RevisionGrid/Columns/GraphColumnProvider.cs#L475-L482

The reason for this is hidden in implementation of `GetLaneRow`:

https://github.com/gitextensions/gitextensions/blob/b210331fb67b2362cc1c666ad49911d4666d85e0/GitUI/UserControls/RevisionGrid/Graph/GraphModel.cs#L26-L44

Whenever the graph is refreshed, the execution goes through `return new SavedLaneRow(_nodes[row]);`.
The reason for that is because `_laneRows` collection is cleared 

```
   at GitUI.UserControls.RevisionGrid.Graph.GraphModel.ClearLanes()
   at GitUI.UserControls.RevisionGrid.Graph.GraphModel.Clear()
   at GitUI.UserControls.RevisionGrid.Columns.GraphColumnProvider.Clear()
   at GitUI.UserControls.RevisionGrid.RevisionDataGridView.Clear()
   at GitUI.RevisionGridControl.ForceRefreshRevisions()
   at GitUI.CommandsDialogs.FormBrowse.RefreshRevisions()
   at GitUI.CommandsDialogs.FormBrowse.RefreshToolStripMenuItemClick(Object sender, EventArgs e)
```

@drewnoakes why do we need to clear models? 
Why do we need to clear the grid at all? Wouldn't it be redrawn during the next refresh event?

If lines 237-238 are commented out the graph rendering is back to expected whenever the graph is refreshed:

https://github.com/gitextensions/gitextensions/blob/b210331fb67b2362cc1c666ad49911d4666d85e0/GitUI/UserControls/RevisionGrid/Graph/GraphModel.cs#L235-L239

I would think the only time we need to clear models is whenever we change a repository or we got new data from `git` command (e.g. fetch, commit etc).
I can't claim to intimately understand the logic, and hence require your input.

**Comment by drewnoakes** [comment]
@RussKie looks sensible. I also find the graph rendering code makes my head hurt a bit. It's probably quite a good implementation, but it's hard to tell what it's doing without much documentation. I have a branch on my fork that attempts to document bits as I work out what they're doing, but it shouldn't block a fix like this. Longer term I'd like to see the large number of blocking `lock` statements removed, but that's a stretch goal.

I don't know why it's clearing. If it works with the change, let's make it, but leave the old code commented with a description of why we did so. It'd be interesting to see whether this actually changed in #5087, as I didn't intentionally make a change in this area. I did, however, move a bunch of the graph rendering code out of a monolith into the graph column/model classes, which may have been the operation that introduced the regression.

**Comment by RussKie** [comment]
I'm observing an art related to stash rendering:
![scroll art1](https://user-images.githubusercontent.com/4403806/46860355-fd311c00-ce5b-11e8-9a07-36d9171d4fe1.gif)

@spdr870 looks like you know the graph rendering better than anyone else, do you think you could look into this one?

**Comment by spdr870** [comment]
I could take a look at it, but how to reproduce?

**Comment by RussKie** [comment]
I'll upload .git folder to a cloud and share with you

**Comment by spdr870** [comment]
That will help. Thanks

Op vr 12 okt. 2018 12:40 schreef RussKie <notifications@github.com>:

> I'll upload .git folder to a cloud and share with you
>

**Comment by RussKie** [comment]
Here's a link to an archived git repo https://drive.google.com/open?id=1vg9v6K_qMFojovPklO_81jjGkhrlPBJw (~550M)

**Comment by spdr870** [comment]
Got the repo, thanks. I did a fix, just do not really like it. I will work on it more.

**Comment by spdr870** [comment]
Fixed in #5610

**Comment by spdr870** [comment]
#5610 has been merged. Closed this issue.
